In [1]:
import random as rn
import numpy as np
from numpy.random import choice as np_choice

# функция для считывания входящих данных
def read_input():
    data = []
    while True:
        try:
            line = input().split()
        except EOFError:
            break
        data.append(line)
    matrix = np.array(data, float)
    return matrix

class AntColony(object):
    def __init__(self, distances, n_ants, n_best, n_iterations, decay, alpha=1, beta=1):
        '''
        # избавляемся от нулей в матрице
        i = 0
        j = 0
        while i < len(distances):
            while j < len(distances):
                if distances[i][j] == 0:
                    distances[i][j] = np.inf
                    i += 1
                    j += 1
                else:
                    continue
        '''            
        self.distances  = distances #матрица растояний
        self.pheromone = np.ones(self.distances.shape) / len(distances) # данные о феромонах
        self.all_inds = range(len(distances)) # список городов
        self.n_ants = n_ants # колличество муравьев
        self.n_best = n_best # колличество элитных муравьев
        self.n_iterations = n_iterations # колличество итераций
        self.decay = decay # испарения феромона
        self.alpha = alpha
        self.beta = beta

    def run(self):
        shortest_path = None
        all_time_shortest_path = ("placeholder", np.inf)
        for i in range(self.n_iterations):
            all_paths = self.gen_all_paths()
            self.spread_pheronome(all_paths, self.n_best, shortest_path=shortest_path)
            shortest_path = min(all_paths, key=lambda x: x[1])
            if shortest_path[1] < all_time_shortest_path[1]:
                all_time_shortest_path = shortest_path
            self.pheromone * self.decay
        return all_time_shortest_path

    def spread_pheronome(self, all_paths, n_best, shortest_path):
        sorted_paths = sorted(all_paths, key=lambda x: x[1])
        for path, dist in sorted_paths[:n_best]:
            for move in path:
                self.pheromone[move] += 1.0 / self.distances[move]

    def gen_path_dist(self, path):
        total_dist = 0
        for ele in path:
            total_dist += self.distances[ele]
        return total_dist

    def gen_all_paths(self):
        all_paths = []
        for i in range(self.n_ants):
            path = self.gen_path(0)
            all_paths.append((path, self.gen_path_dist(path)))
        return all_paths

    def gen_path(self, start):
        path = []
        visited = set()
        visited.add(start)
        prev = start
        for i in range(len(self.distances) - 1):
            move = self.pick_move(self.pheromone[prev], self.distances[prev], visited)
            path.append((prev, move))
            prev = move
            visited.add(move)
        path.append((prev, start)) # going back to where we started
        return path

    def pick_move(self, pheromone, dist, visited):
        pheromone = np.copy(pheromone)
        pheromone[list(visited)] = 0
        row = pheromone ** self.alpha * (( 1.0 / dist) ** self.beta)
        norm_row = row / row.sum()
        move = np_choice(self.all_inds, 1, p=norm_row)[0]
        return move



In [2]:
#distance=read_input()
#distance depends on direction
from pandas import read_csv
dataset = read_csv('data.csv',';')
dataset.head(7)

x1 
 x2 
 x3 
 x4 
 x5 
 x6 
 
 
 
 
 0 
 88.4 
 98.1 
 73.3 
 35.5 
 5.1 
 30.4 
 
 
 1 
 70.6 
 50.6 
 51.2 
 13.9 
 29.0 
 8.4 
 
 
 2 
 44.7 
 91.9 
 88.9 
 1.1 
 10.0 
 20.6 
 
 
 3 
 56.0 
 49.1 
 84.5 
 79.4 
 53.3 
 75.8 
 
 
 4 
 85.8 
 24.8 
 34.0 
 42.8 
 5.5 
 86.3 
 
 
 5 
 70.8 
 6.5 
 63.4 
 91.2 
 48.1 
 63.3

In [3]:
distance=dataset.values
#print(distance[2][4])
ant_colony1 = AntColony(distance, 10,5,10, 0.95, alpha=1, beta=1)
shortest_path = ant_colony1.run()
print (int(shortest_path[1]))
        

158


In [18]:
ant_colony2 = AntColony(distance, 20,5,10, 0.95, alpha=1, beta=1)
shortest_path = ant_colony2.run()
print (int(shortest_path[1]))
        

136


In [22]:
ant_colony3 = AntColony(distance, 20,10,10, 0.95, alpha=1, beta=1)
shortest_path = ant_colony3.run()
print (int(shortest_path[1]))
        

136


In [6]:
ant_colony4 = AntColony(distance, 20,10,10, 0.95, alpha=2, beta=1)
shortest_path = ant_colony4.run()
print (int(shortest_path[1]))

158


In [7]:
ant_colony5 = AntColony(distance, 20,10,10, 0.95, alpha=2, beta=2)
shortest_path = ant_colony5.run()
print (int(shortest_path[1]))

158


In [8]:
ant_colony6 = AntColony(distance, 20,10,1000, 0.95, alpha=2, beta=2)
shortest_path = ant_colony6.run()
print (int(shortest_path[1]))

158


In [23]:
ant_colony7 = AntColony(distance, 20,5,1000, 0.95, alpha=2, beta=2)
shortest_path = ant_colony7.run()
print (int(shortest_path[1]))

158


In [10]:
ant_colony8 = AntColony(distance, 20,5,1000, 0.95, alpha=2, beta=1)
shortest_path = ant_colony8.run()
print (int(shortest_path[1]))

136


In [11]:
ant_colony9 = AntColony(distance, len(distance)*2,5,len(distance)*4, 0.95, alpha=1, beta=1)
shortest_path = ant_colony9.run()
print (int(shortest_path[1]))

158


In [12]:
ant_colony10 = AntColony(distance, len(distance)*2,5,len(distance)*4, 0.8, alpha=1, beta=1)
shortest_path = ant_colony10.run()
print (int(shortest_path[1]))

158


In [13]:
ant_colony11 = AntColony(distance, len(distance)*2,5,len(distance)*4, 0.8, alpha=2, beta=1)
shortest_path = ant_colony11.run()
print (int(shortest_path[1]))

158


In [14]:
ant_colony8 = AntColony(distance, 100,20,1000, 0.95, alpha=2, beta=1)
shortest_path = ant_colony8.run()
print (int(shortest_path[1]))

157


Чаще всего выводились 158 и 157. С одинаковой частотой для разных наборов парметров. Лучший результат показывался -136. Только с показателем beta=1. Увеличение итераций без увеличения количества агентов (муравьёв) показатели не улучшили. Как и случай, когда количество итераций и муравьёв во много раз препосходит размеры матрицы.
Наиболее оптимальные параметры (по частоте вывода лучшего результата - 136) : AntColony(distance, 20,10,10, 0.95, alpha=1, beta=1). То есть n_ants=20, n_best=10, n_iterations=10, decay=0.95, alpha=1, beta=1

In [15]:
dataset_simm = read_csv('data_simm.csv',';')
dataset_simm.head(7)

x1 
 x2 
 x3 
 x4 
 x5 
 x6 
 
 
 
 
 0 
 88.4 
 98.1 
 73.3 
 35.5 
 5.1 
 30.4 
 
 
 1 
 98.1 
 50.6 
 51.2 
 13.9 
 29.0 
 8.4 
 
 
 2 
 73.3 
 51.2 
 88.9 
 1.1 
 10.0 
 20.6 
 
 
 3 
 35.5 
 13.9 
 1.1 
 79.4 
 53.3 
 75.8 
 
 
 4 
 5.1 
 29.0 
 10.0 
 53.3 
 5.5 
 86.3 
 
 
 5 
 30.4 
 8.4 
 20.6 
 75.8 
 86.3 
 63.3

In [16]:
distance_simm=dataset_simm.values
ant_colony_simm1 = AntColony(distance, 10,5,10, 0.95, alpha=1, beta=1)
shortest_path = ant_colony_simm1.run()
print (int(shortest_path[1]))

158


In [17]:
distance_simm=dataset_simm.values
ant_colony_simm2 = AntColony(distance, 20,5,10, 0.95, alpha=1, beta=1)
shortest_path = ant_colony_simm2.run()
print (int(shortest_path[1]))

158
